django/template/defaulttags.py

```diff
@@ -1564,3 +1567,186 @@ def do_with(parser, token):
     nodelist = parser.parse(("endwith",))
     parser.delete_first_token()
     return WithNode(None, None, nodelist, extra_context=extra_context)
+
+
+class TemplateProxy:
+    """
+    A lightweight Template lookalike used for template partials.
+
+    Wraps nodelist as partial, in order to bind context.
+    """
+
+    def __init__(self, nodelist, origin, name):
+        self.nodelist = nodelist
+        self.origin = origin
+        self.name = name
+
+    def get_exception_info(self, exception, token):
+        template = self.origin.loader.get_template(self.origin.template_name)
+        return template.get_exception_info(exception, token)
+
+    def find_partial_source(self, full_source, partial_name):
+        result = ""
+        pos = 0
+        for m in partial_start_tag_re.finditer(full_source, pos):
+            sspos, sepos = m.span()
+            starter, name, inline = m.groups()
+            endm = partial_end_tag_re.search(full_source, sepos + 1)
+            assert endm, "End tag must be present"
+            espos, eepos = endm.span()
+            if name == partial_name:
+                # Include the full partial definition from opening to closing tag
+                result = full_source[sspos:eepos]
+                break
+            pos = eepos + 1
+        return result
+
+    @property
+    def source(self):
+        template = self.origin.loader.get_template(self.origin.template_name)
+        return self.find_partial_source(template.source, self.name)
+
+    def _render(self, context):
+        return self.nodelist.render(context)
+
+    def render(self, context):
+        "Display stage -- can be called many times"
+        with context.render_context.push_state(self):
+            if context.template is None:
+                with context.bind_template(self):
+                    context.template_name = self.name
+                    return self._render(context)
+            else:
+                return self._render(context)
+
+
+class DefinePartialNode(Node):
+    def __init__(self, partial_name, inline, nodelist):
+        self.partial_name = partial_name
+        self.inline = inline
+        self.nodelist = nodelist
+
+    def render(self, context):
+        """Set content into context and return empty string"""
```

Comment:
This docstring isn't right. (Historically it was, but then we added the `inline` option, and it should have been corrected then. 

I'd might remove it at this point.